### PR TITLE
fix(android/engine): Make it easier to test keyboard updates

### PR DESCRIPTION
**Pull request by darcywong00**
Currently there's a lot of hoops involved for testing Keyman keyboard package updates (see https://github.com/keymanapp/keyman/pull/7740#issuecomment-1319640048)

This PR adds overrides for local and PR test builds to ignore calendar comparisons for checking keyboard updates. (Note: this won't apply to release builds used in Acceptance Testing)

I also separated the updateKMP comparison so that
* keyboard info is updated if the cloud has a newer version
* if updateKMP has a value, trigger notification for update available (this was causing updates to only appear on the 2nd boot)

Additional changes:
* Refactors version environment checks to new `VersionUtils()`

Commit https://github.com/keymanapp/keyman/commit/2d0008e2e38e1ffc0481af5f84a5f88bd8a1e3b7 is temporary test code for installing an old version 6.0.4 of sil_cameroon_qwerty.
- [x] DO NOT MERGE until reverting test code

## User Testing
Setup - Use Android device/emulator (phone in portrait orientation) with network access enabled

* **TEST_KEYBOARD_UPATES** 
1. Install the PR build of Keyman for Android
2. Launch Keyman for Android a first time.  This installs the old version 6.0.4 of sil_cameroon_qwerty
3. Completely close Keyman
4. Launch Keyman for Android a second time. This performs a check for keyboard updates
5. Dismiss the "Get Started" menu
6. Click the Keyman overflow menu
7. Verify icon for "Install Updates" is visible

### CODE REVIEW

File: android/KMAPro/kMAPro/src/main/java/com/keyman/android/CheckInstallReferrer.java [hunk 1]
```diff
@@ -66,8 +66,7 @@ public static void checkGooglePlayInstallReferrer(MainActivity mainActivity, Con
     editor.commit();
 
     // local environment or test builds are nearly always side loaded
-    if (BuildConfig.VERSION_ENVIRONMENT.equalsIgnoreCase("local") ||
-        BuildConfig.VERSION_NAME.matches("^.*(-test-\\d+)$")) {
+    if (VersionUtils.isLocalBuild() || VersionUtils.isTestBuild()) {
```
Comment (mcdurdin) [inline android/KMAPro/kMAPro/src/main/java/com/keyman/android/CheckInstallReferrer.java:69]: :+1:

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/cloud/CloudDataJsonUtil.java [hunk 2]
```diff
@@ -153,9 +153,10 @@ public static void processKeyboardPackageUpdateJSON(Context aContext, JSONObject
             for (int i = 0; i < KeyboardController.getInstance().get().size(); i++) {
               Keyboard kbd = KeyboardController.getInstance().getKeyboardInfo(i);
               String version = kbd.getVersion();
+              String updateKMP = kbd.getUpdateKMP();
               if (keyboardID.equalsIgnoreCase(kbd.getKeyboardID()) &&
                   (FileUtils.compareVersions(cloudVersion, version) == FileUtils.VERSION_GREATER) &&
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/cloud/CloudDataJsonUtil.java:158]:
```suggestion
                  FileUtils.compareVersions(cloudVersion, version) == FileUtils.VERSION_GREATER &&
```

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/cloud/CloudDataJsonUtil.java [hunk 3]
```diff
@@ -153,9 +153,10 @@ public static void processKeyboardPackageUpdateJSON(Context aContext, JSONObject
...
-                  (!kbd.getUpdateKMP().equalsIgnoreCase(cloudKMP))) {
+                  (updateKMP != null && !updateKMP.equalsIgnoreCase(cloudKMP))) {
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/cloud/CloudDataJsonUtil.java:159]: Does this fix a known crash?

Suggestion is just cleanup - remove unnecessary parentheses for easier reading of test clauses:

```suggestion
                  updateKMP != null && 
                  !updateKMP.equalsIgnoreCase(cloudKMP)) {
```

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudRepository.java [hunk 4]
```diff
@@ -89,6 +91,10 @@ public boolean getCacheValidity(@NonNull Context context) {
     boolean loadResourcesFromCache = this.shouldUseCache(context, CloudDataJsonUtil.getResourcesCacheFile(context));
 
     boolean cacheValid = loadLexicalModelsFromCache && loadResourcesFromCache;
+
+    if (VersionUtils.isLocalBuild() || VersionUtils.isTestBuild()) {
+      return false;
+    }
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudRepository.java:97]: This should be at the top of the function with a comment explaining why we are doing this.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/logic/ResourcesUpdateTool.java [hunk 5]
```diff
@@ -464,6 +470,9 @@ public void run() {
   }
 
   public boolean shouldCheckUpdate(Context aContext) {
+    if (VersionUtils.isLocalBuild() || VersionUtils.isTestBuild() || FORCE_RESOURCE_UPDATE) {
+      return true;
+    }
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/logic/ResourcesUpdateTool.java:475]: Please add a comment explaining why we return true in this case :grin:

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/VersionUtils.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+public final class VersionUtils {
+  public static final String TAG = "VersionUtils";
+
+  public static boolean isLocalBuild() {
+    return BuildConfig.VERSION_ENVIRONMENT.equalsIgnoreCase("local");
+  }
+
+  public static boolean isTestBuild() {
+    return BuildConfig.KEYMAN_ENGINE_VERSION_NAME.matches("^.*(-test-\\d+)$");
+  }
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/VersionUtils.java:20]: Won't `VERSION_ENVIRONMENT` be "test" in this case?

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/cloud/CloudDataJsonUtil.java (hunk 3 above)
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/cloud/CloudDataJsonUtil.java:159]: > Does this fix a known crash?

Pretty [rare one](https://sentry.io/organizations/keyman/issues/3659638286/?project=5983520) 
1 user on 15.0.270 stable and myself

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/VersionUtils.java (hunk 6 above)
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/VersionUtils.java:20]: oh yeah, that would be simpler.

**Comment by darcywong00** [comment]
So afaict, all the test code is here. And I consistently get update notifications when Keyman is launched the second time. But after closing Keyman, I can't get the notifications to reappear with Keyman :/

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/cloud/CloudDataJsonUtil.java (hunk 2 above)
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/cloud/CloudDataJsonUtil.java]: I'll isolate this change to a separate PR #7787

**Comment by bharanidharanj** [comment]
* **TEST_KEYBOARD_UPATES (PASSED):** Tested with the attached PR build in Android 9.0 emulator and here is my observation: 1. Able to see "sil_cameroon_qwerty 6.0.4" version in Installed Languages menu. 2. Closed the Keyman app. Re-opened it. 3. After few minutes, I just clicked the Keyman menu list and noticed that "Install Updates" option appears with other options.

**Comment by darcywong00** [comment]
User testing passed, so I've reverted the test code in https://github.com/keymanapp/keyman/pull/7781/commits/29bd8dce09765750063e7ecb68288c8e0dfea18b

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudRepository.java [hunk 7]
```diff
@@ -61,6 +62,11 @@ private CloudRepository() {
     lastLoad = null;
   }
 
+  // For local and PR test builds, force check of keyboard updates
+  private static boolean isTestBuild() {
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudRepository.java:66]:
```suggestion
  private static boolean isLocalOrTestBuild() {
```

Given you are using the same pattern also in other source files, perhaps you should move this to VersionUtils.java and update references accordingly?

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudRepository.java [hunk 8]
```diff
@@ -238,7 +245,8 @@ public void updateDatasetIfNeeded(@NonNull Context context, UpdateHandler update
   {
     boolean cacheValid = getCacheValidity(context);
 
-    if(cacheValid && shouldUseMemCache(context)) {
+    // For local and PR test builds, force update dataset
+    if(cacheValid && shouldUseMemCache(context) && !isTestBuild()) {
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudRepository.java:249]:
```suggestion
    if(cacheValid && shouldUseMemCache(context) && !isLocalOrTestBuild()) {
```

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudRepository.java [hunk 9]
```diff
@@ -383,7 +392,8 @@ private void downloadMetaDataFromServer(@NonNull Context context, UpdateHandler
     //    int cloudQueryEntries = 0;
     List<CloudApiTypes.CloudApiParam> cloudQueries = new ArrayList<>(2);
 
-    if (!cacheValid) {
+    // For local and PR test builds, force check of keyboard updates
+    if (!cacheValid || isTestBuild()) {
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudRepository.java:396]:
```suggestion
    if (!cacheValid || isLocalOrTestBuild()) {
```

**Review by mcdurdin — CHANGES_REQUESTED** [review]
Looking close but a minor refactor requested

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudRepository.java (hunk 7 above)
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudRepository.java:66]: Moving to `VersionUtils.isLocalOrTestBuild()`

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/cloud/impl/CloudCatalogDownloadCallback.java [hunk 10]
```diff
@@ -69,7 +70,8 @@ public void onUpdateDetection(List<Bundle> updateBundles) {
     };
   }
   private Bundle updateCheck(LanguageResource cloudResource, LanguageResource existingMatch) {
-    if (DEBUG_SIMULATE_UPDATES) {
+    // For local and PR test builds, invalidate cache to make keyboard updates easier
+    if (DEBUG_SIMULATE_UPDATES || VersionUtils.isLocalBuild() || VersionUtils.isTestBuild()) {
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/cloud/impl/CloudCatalogDownloadCallback.java:74]:
```suggestion
    if (DEBUG_SIMULATE_UPDATES || VersionUtils.isLocalOrTestBuild()) {
```

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/logic/ResourcesUpdateTool.java [hunk 11]
```diff
@@ -464,6 +470,10 @@ public void run() {
   }
 
   public boolean shouldCheckUpdate(Context aContext) {
+    // For local and PR test builds, invalidate cache to make keyboard updates easier
+    if (VersionUtils.isLocalBuild() || VersionUtils.isTestBuild() || FORCE_RESOURCE_UPDATE) {
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/logic/ResourcesUpdateTool.java:474]:
```suggestion
    if (VersionUtils.isLocalOrTestBuild() || FORCE_RESOURCE_UPDATE) {
```

**Review by mcdurdin — APPROVED** [review]

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 16.0.111-beta](https://keyman.com/downloads/releases/beta/16.0.111)

**Pull request merged**